From 6c6dc11922d859006375972b554d21e733bc086d Mon Sep 17 00:00:00 2001
From: xzhang <xzhang@xzdt>
Date: Tue, 28 Jul 2015 11:39:37 -0700
Subject: [PATCH 0425/1164] CDH-29554: Backport: HIVE-11288: Avro SerDe
 InstanceCache returns incorrect schema (Greg via
 Xuefu)

Conflicts:
	serde/src/java/org/apache/hadoop/hive/serde2/avro/InstanceCache.java
	serde/src/test/org/apache/hadoop/hive/serde2/avro/TestInstanceCache.java

Conflicts:

	serde/src/java/org/apache/hadoop/hive/serde2/avro/InstanceCache.java

Change-Id: Ia7ad3d0591e561e10d47845fa3582f82723cb166
---
 .../hadoop/hive/serde2/avro/InstanceCache.java     |    9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/serde/src/java/org/apache/hadoop/hive/serde2/avro/InstanceCache.java b/serde/src/java/org/apache/hadoop/hive/serde2/avro/InstanceCache.java
index 0b10d99..c7289cb 100644
--- a/serde/src/java/org/apache/hadoop/hive/serde2/avro/InstanceCache.java
+++ b/serde/src/java/org/apache/hadoop/hive/serde2/avro/InstanceCache.java
@@ -21,6 +21,7 @@
 import org.apache.commons.logging.LogFactory;
 
 import java.util.HashMap;
+import java.util.Map;
 import java.util.Set;
 
 /**
@@ -33,7 +34,7 @@
  */
 public abstract class InstanceCache<SeedObject, Instance> {
   private static final Log LOG = LogFactory.getLog(InstanceCache.class);
-  HashMap<Integer, Instance> cache = new HashMap<Integer, Instance>();
+  Map<SeedObject, Instance> cache = new HashMap<SeedObject, Instance>();
   
   public InstanceCache() {}
 
@@ -53,15 +54,15 @@ public Instance retrieve(SeedObject hv,
       Set<SeedObject> seenSchemas) throws AvroSerdeException {
     if(LOG.isDebugEnabled()) LOG.debug("Checking for hv: " + hv.toString());
 
-    if(cache.containsKey(hv.hashCode())) {
+    if(cache.containsKey(hv)) {
       if(LOG.isDebugEnabled()) LOG.debug("Returning cache result.");
-      return cache.get(hv.hashCode());
+      return cache.get(hv);
     }
 
     if(LOG.isDebugEnabled()) LOG.debug("Creating new instance and storing in cache");
 
     Instance instance = makeInstance(hv, seenSchemas);
-    cache.put(hv.hashCode(), instance);
+    cache.put(hv, instance);
     return instance;
   }
 
-- 
1.7.9.5

